### DISCUSSION TITLE
Best practice of using pgtype package

### DISCUSSION THREAD

**Discussion by hnsiri** [jackc/pgx#1853] [Q&A] [answered]
Hi all,

I have a question about the `pgtype` package. When should we use it? Is it good practice to load data from the database into these types and then assign them to Go types? I have two structs; one is the `dbSchema` with all fields using "pgtype" types, and the second one is an entity with built-in types. Whenever I read data from the database, I first scan them into `dbSchema` and then load them into my entity.

What do you think about it? Any other ideas or suggestions?

**Comment by jackc** [comment, marked as answer]
In general, you should only use pgtype directly when you need to. For example, if you have a `NOT NULL integer` in PostgreSQL then scan it directly to a Go `int32`. But if it is nullable then scan it into a `pgtype.Int4`. 

You can also have your types implement pgtype interfaces like `DateScanner` and `DateValuer`. That allows you to directly use your application types.

So in short, try to use native Go types or your application types and only use pgtype types directly when necessary.

**Reply by hnsiri** [reply]
Great! I appreciate your guidance. Thank you.
